File: clients/go/consensus/parse.go
package consensus

import "fmt"

func parseInput(cur *cursor) (TxInput, error) {
	prevTxidBytes, err := cur.readExact(32)
	if err != nil {
		return TxInput{}, err
	}
	var prevTxid [32]byte
	copy(prevTxid[:], prevTxidBytes)

	prevVout, err := cur.readU32LE()
	if err != nil {
		return TxInput{}, err
	}

	scriptSigLenU64, err := cur.readCompactSize()
	if err != nil {
		return TxInput{}, err
	}
	scriptSigLen, err := toIntLen(scriptSigLenU64, "script_sig_len")
	if err != nil {
		return TxInput{}, err
	}
	scriptSigBytes, err := cur.readExact(scriptSigLen)
	if err != nil {
		return TxInput{}, err
	}

	sequence, err := cur.readU32LE()
	if err != nil {
		return TxInput{}, err
	}

	return TxInput{
		PrevTxid:  prevTxid,
		PrevVout:  prevVout,
		ScriptSig: append([]byte(nil), scriptSigBytes...),
		Sequence:  sequence,
	}, nil
}

func parseOutput(cur *cursor) (TxOutput, error) {
	value, err := cur.readU64LE()
	if err != nil {
		return TxOutput{}, err
	}
	covenantType, err := cur.readU16LE()
	if err != nil {
		return TxOutput{}, err
	}

	covenantDataLenU64, err := cur.readCompactSize()
	if err != nil {
		return TxOutput{}, err
	}
	covenantDataLen, err := toIntLen(covenantDataLenU64, "covenant_data_len")
	if err != nil {
		return TxOutput{}, err
	}
	covenantDataBytes, err := cur.readExact(covenantDataLen)
	if err != nil {
		return TxOutput{}, err
	}

	return TxOutput{
		Value:        value,
		CovenantType: covenantType,
		CovenantData: append([]byte(nil), covenantDataBytes...),
	}, nil
}

func parseWitnessItem(cur *cursor) (WitnessItem, error) {
	suiteID, err := cur.readU8()
	if err != nil {
		return WitnessItem{}, err
	}

	pubkeyLenU64, err := cur.readCompactSize()
	if err != nil {
		return WitnessItem{}, err
	}
	pubkeyLen, err := toIntLen(pubkeyLenU64, "pubkey_len")
	if err != nil {
		return WitnessItem{}, err
	}
	pubkeyBytes, err := cur.readExact(pubkeyLen)
	if err != nil {
		return WitnessItem{}, err
	}

	sigLenU64, err := cur.readCompactSize()
	if err != nil {
		return WitnessItem{}, err
	}
	sigLen, err := toIntLen(sigLenU64, "sig_len")
	if err != nil {
		return WitnessItem{}, err
	}
	sigBytes, err := cur.readExact(sigLen)
	if err != nil {
		return WitnessItem{}, err
	}

	return WitnessItem{
		SuiteID:   suiteID,
		Pubkey:    append([]byte(nil), pubkeyBytes...),
		Signature: append([]byte(nil), sigBytes...),
	}, nil
}

func parseInputList(cur *cursor) ([]TxInput, error) {
	inputCountU64, err := cur.readCompactSize()
	if err != nil {
		return nil, err
	}
	inputCount, err := toIntLen(inputCountU64, "input_count")
	if err != nil {
		return nil, err
	}
	inputs := make([]TxInput, 0, inputCount)
	for i := 0; i < inputCount; i++ {
		inp, err := parseInput(cur)
		if err != nil {
			return nil, err
		}
		inputs = append(inputs, inp)
	}
	return inputs, nil
}

func parseOutputList(cur *cursor) ([]TxOutput, error) {
	outputCountU64, err := cur.readCompactSize()
	if err != nil {
		return nil, err
	}
	outputCount, err := toIntLen(outputCountU64, "output_count")
	if err != nil {
		return nil, err
	}
	outputs := make([]TxOutput, 0, outputCount)
	for i := 0; i < outputCount; i++ {
		out, err := parseOutput(cur)
		if err != nil {
			return nil, err
		}
		outputs = append(outputs, out)
	}
	return outputs, nil
}

func parseWitnessList(cur *cursor) ([]WitnessItem, error) {
	witnessCountU64, err := cur.readCompactSize()
	if err != nil {
		return nil, err
	}
	witnessCount, err := toIntLen(witnessCountU64, "witness_count")
	if err != nil {
		return nil, err
	}
	witnesses := make([]WitnessItem, 0, witnessCount)
	for i := 0; i < witnessCount; i++ {
		w, err := parseWitnessItem(cur)
		if err != nil {
			return nil, err
		}
		witnesses = append(witnesses, w)
	}
	return witnesses, nil
}

func ParseTxBytes(b []byte) (*Tx, error) {
	cur := newCursor(b)
	tx, err := ParseTxBytesFromCursor(cur)
	if err != nil {
		return nil, err
	}
	if cur.pos != len(b) {
		return nil, fmt.Errorf("parse: trailing bytes")
	}
	return tx, nil
}

// ParseTxBytesPrefix parses a single transaction from the start of b and returns
// the number of bytes consumed. This is used by P2P compact-block messages where
// multiple TxBytes are concatenated without explicit per-tx length prefixes.
func ParseTxBytesPrefix(b []byte) (*Tx, int, error) {
	cur := newCursor(b)
	tx, err := ParseTxBytesFromCursor(cur)
	if err != nil {
		return nil, 0, err
	}
	return tx, cur.pos, nil
}

func ParseBlockHeader(cur *cursor) (BlockHeader, error) {
	version, err := cur.readU32LE()
	if err != nil {
		return BlockHeader{}, err
	}
	prev, err := cur.readExact(32)
	if err != nil {
		return BlockHeader{}, err
	}
	merkle, err := cur.readExact(32)
	if err != nil {
		return BlockHeader{}, err
	}
	timestamp, err := cur.readU64LE()
	if err != nil {
		return BlockHeader{}, err
	}
	target, err := cur.readExact(32)
	if err != nil {
		return BlockHeader{}, err
	}
	nonce, err := cur.readU64LE()
	if err != nil {
		return BlockHeader{}, err
	}
	var target32 [32]byte
	copy(target32[:], target)
	var prev32 [32]byte
	copy(prev32[:], prev)
	var merkle32 [32]byte
	copy(merkle32[:], merkle)
	return BlockHeader{
		Version:       version,
		PrevBlockHash: prev32,
		MerkleRoot:    merkle32,
		Timestamp:     timestamp,
		Target:        target32,
		Nonce:         nonce,
	}, nil
}

func ParseBlockBytes(b []byte) (Block, error) {
	cur := newCursor(b)
	header, err := ParseBlockHeader(cur)
	if err != nil {
		return Block{}, err
	}
	txCountU64, err := cur.readCompactSize()
	if err != nil {
		return Block{}, err
	}
	txCount, err := toIntLen(txCountU64, "tx_count")
	if err != nil {
		return Block{}, err
	}
	txs := make([]Tx, 0, txCount)
	for i := 0; i < txCount; i++ {
		tx, err := ParseTxBytesFromCursor(cur)
		if err != nil {
			return Block{}, err
		}
		txs = append(txs, *tx)
	}
	if cur.pos != len(b) {
		return Block{}, fmt.Errorf("BLOCK_ERR_PARSE")
	}
	return Block{
		Header:       header,
		Transactions: txs,
	}, nil
}

func ParseTxBytesFromCursor(cur *cursor) (*Tx, error) {
	version, err := cur.readU32LE()
	if err != nil {
		return nil, err
	}
	if version != TX_VERSION_V2 {
		return nil, fmt.Errorf("TX_ERR_PARSE")
	}
	kindU8, err := cur.readU8()
	if err != nil {
		return nil, err
	}
	txKind := uint8(kindU8)
	txNonce, err := cur.readU64LE()
	if err != nil {
		return nil, err
	}
	inputs, err := parseInputList(cur)
	if err != nil {
		return nil, err
	}
	outputs, err := parseOutputList(cur)
	if err != nil {
		return nil, err
	}
	locktime, err := cur.readU32LE()
	if err != nil {
		return nil, err
	}

	var daCommit *DACommitFields
	var daChunk *DAChunkFields
	switch txKind {
	case TX_KIND_STANDARD:
		// no DA core fields
	case TX_KIND_DA_COMMIT:
		var f DACommitFields
		daid, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.DAID[:], daid)
		cc, err := cur.readU16LE()
		if err != nil {
			return nil, err
		}
		f.ChunkCount = cc
		domain, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.RETLDomainID[:], domain)
		bn, err := cur.readU64LE()
		if err != nil {
			return nil, err
		}
		f.BatchNumber = bn
		r1, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.TxDataRoot[:], r1)
		r2, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.StateRoot[:], r2)
		r3, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.WithdrawalsRoot[:], r3)
		suite, err := cur.readU8()
		if err != nil {
			return nil, err
		}
		f.BatchSigSuite = uint8(suite)
		sigLenU64, err := cur.readCompactSize()
		if err != nil {
			return nil, err
		}
		sigLen, err := toIntLen(sigLenU64, "batch_sig_len")
		if err != nil {
			return nil, err
		}
		sig, err := cur.readExact(sigLen)
		if err != nil {
			return nil, err
		}
		f.BatchSig = append([]byte(nil), sig...)
		daCommit = &f
	case TX_KIND_DA_CHUNK:
		var f DAChunkFields
		daid, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.DAID[:], daid)
		ci, err := cur.readU16LE()
		if err != nil {
			return nil, err
		}
		f.ChunkIndex = ci
		h, err := cur.readExact(32)
		if err != nil {
			return nil, err
		}
		copy(f.ChunkHash[:], h)
		daChunk = &f
	default:
		return nil, fmt.Errorf("TX_ERR_PARSE")
	}

	witnesses, err := parseWitnessList(cur)
	if err != nil {
		return nil, err
	}

	daPayloadLenU64, err := cur.readCompactSize()
	if err != nil {
		return nil, err
	}
	daPayloadLen, err := toIntLen(daPayloadLenU64, "da_payload_len")
	if err != nil {
		return nil, err
	}
	daPayload, err := cur.readExact(daPayloadLen)
	if err != nil {
		return nil, err
	}

	if txKind == TX_KIND_STANDARD {
		if daPayloadLen != 0 {
			return nil, fmt.Errorf("TX_ERR_PARSE")
		}
	} else {
		if daPayloadLen == 0 {
			return nil, fmt.Errorf("TX_ERR_PARSE")
		}
		// Per-tx consensus caps.
		if txKind == TX_KIND_DA_COMMIT && daPayloadLen > MAX_DA_MANIFEST_BYTES_PER_TX {
			return nil, fmt.Errorf("TX_ERR_PARSE")
		}
		if txKind == TX_KIND_DA_CHUNK && daPayloadLen > MAX_DA_CHUNK_BYTES_PER_TX {
			return nil, fmt.Errorf("TX_ERR_PARSE")
		}
	}

	return &Tx{
		Version:   version,
		TxKind:    txKind,
		TxNonce:   txNonce,
		Inputs:    inputs,
		Outputs:   outputs,
		Locktime:  locktime,
		DACommit:  daCommit,
		DAChunk:   daChunk,
		DAPayload: append([]byte(nil), daPayload...),
		Witness:   WitnessSection{Witnesses: witnesses},
	}, nil
}

File: clients/rust/crates/rubin-consensus/src/parse.rs
use crate::wire::Cursor;
use crate::{
    BLOCK_ERR_PARSE, Block, BlockHeader, DAChunkFields, DACommitFields, MAX_DA_CHUNK_BYTES_PER_TX,
    MAX_DA_MANIFEST_BYTES_PER_TX, TX_KIND_DA_CHUNK, TX_KIND_DA_COMMIT, TX_KIND_STANDARD,
    TX_VERSION_V2, Tx, TxInput, TxOutput, WitnessItem, WitnessSection,
};

pub fn parse_tx_bytes(bytes: &[u8]) -> Result<Tx, String> {
    let mut cursor = Cursor::new(bytes);
    let tx = parse_tx_from_cursor(&mut cursor)?;
    if cursor.pos != bytes.len() {
        return Err("parse: trailing bytes".into());
    }
    Ok(tx)
}

pub(crate) fn parse_tx_from_cursor(cursor: &mut Cursor<'_>) -> Result<Tx, String> {
    let version = cursor.read_u32le()?;
    if version != TX_VERSION_V2 {
        return Err("TX_ERR_PARSE".into());
    }
    let tx_kind = cursor.read_u8()?;
    let tx_nonce = cursor.read_u64le()?;

    let input_count_u64 = cursor.read_compact_size()?;
    let input_count: usize = input_count_u64
        .try_into()
        .map_err(|_| "parse: input_count overflows usize".to_string())?;
    let mut inputs = Vec::with_capacity(input_count);
    for _ in 0..input_count {
        let prev_txid_slice = cursor.read_exact(32)?;
        let mut prev_txid = [0u8; 32];
        prev_txid.copy_from_slice(prev_txid_slice);
        let prev_vout = cursor.read_u32le()?;

        let script_sig_len_u64 = cursor.read_compact_size()?;
        let script_sig_len: usize = script_sig_len_u64
            .try_into()
            .map_err(|_| "parse: script_sig_len overflows usize".to_string())?;
        let script_sig = cursor.read_exact(script_sig_len)?.to_vec();
        let sequence = cursor.read_u32le()?;

        inputs.push(TxInput {
            prev_txid,
            prev_vout,
            script_sig,
            sequence,
        });
    }

    let output_count_u64 = cursor.read_compact_size()?;
    let output_count: usize = output_count_u64
        .try_into()
        .map_err(|_| "parse: output_count overflows usize".to_string())?;
    let mut outputs = Vec::with_capacity(output_count);
    for _ in 0..output_count {
        let value = cursor.read_u64le()?;
        let covenant_type = cursor.read_u16le()?;

        let covenant_data_len_u64 = cursor.read_compact_size()?;
        let covenant_data_len: usize = covenant_data_len_u64
            .try_into()
            .map_err(|_| "parse: covenant_data_len overflows usize".to_string())?;
        let covenant_data = cursor.read_exact(covenant_data_len)?.to_vec();

        outputs.push(TxOutput {
            value,
            covenant_type,
            covenant_data,
        });
    }

    let locktime = cursor.read_u32le()?;

    let mut da_commit: Option<DACommitFields> = None;
    let mut da_chunk: Option<DAChunkFields> = None;
    match tx_kind {
        TX_KIND_STANDARD => {}
        TX_KIND_DA_COMMIT => {
            let da_id_slice = cursor.read_exact(32)?;
            let mut da_id = [0u8; 32];
            da_id.copy_from_slice(da_id_slice);
            let chunk_count = cursor.read_u16le()?;
            let retl_domain_id_slice = cursor.read_exact(32)?;
            let mut retl_domain_id = [0u8; 32];
            retl_domain_id.copy_from_slice(retl_domain_id_slice);
            let batch_number = cursor.read_u64le()?;
            let tx_data_root_slice = cursor.read_exact(32)?;
            let mut tx_data_root = [0u8; 32];
            tx_data_root.copy_from_slice(tx_data_root_slice);
            let state_root_slice = cursor.read_exact(32)?;
            let mut state_root = [0u8; 32];
            state_root.copy_from_slice(state_root_slice);
            let withdrawals_root_slice = cursor.read_exact(32)?;
            let mut withdrawals_root = [0u8; 32];
            withdrawals_root.copy_from_slice(withdrawals_root_slice);
            let batch_sig_suite = cursor.read_u8()?;
            let batch_sig_len_u64 = cursor.read_compact_size()?;
            let batch_sig_len: usize = batch_sig_len_u64
                .try_into()
                .map_err(|_| "parse: batch_sig_len overflows usize".to_string())?;
            let batch_sig = cursor.read_exact(batch_sig_len)?.to_vec();
            da_commit = Some(DACommitFields {
                da_id,
                chunk_count,
                retl_domain_id,
                batch_number,
                tx_data_root,
                state_root,
                withdrawals_root,
                batch_sig_suite,
                batch_sig,
            });
        }
        TX_KIND_DA_CHUNK => {
            let da_id_slice = cursor.read_exact(32)?;
            let mut da_id = [0u8; 32];
            da_id.copy_from_slice(da_id_slice);
            let chunk_index = cursor.read_u16le()?;
            let chunk_hash_slice = cursor.read_exact(32)?;
            let mut chunk_hash = [0u8; 32];
            chunk_hash.copy_from_slice(chunk_hash_slice);
            da_chunk = Some(DAChunkFields {
                da_id,
                chunk_index,
                chunk_hash,
            });
        }
        _ => return Err("TX_ERR_PARSE".into()),
    }

    let witness_count_u64 = cursor.read_compact_size()?;
    let witness_count: usize = witness_count_u64
        .try_into()
        .map_err(|_| "parse: witness_count overflows usize".to_string())?;
    let mut witnesses = Vec::with_capacity(witness_count);
    for _ in 0..witness_count {
        let suite_id = cursor.read_u8()?;

        let pubkey_len_u64 = cursor.read_compact_size()?;
        let pubkey_len: usize = pubkey_len_u64
            .try_into()
            .map_err(|_| "parse: pubkey_len overflows usize".to_string())?;
        let pubkey = cursor.read_exact(pubkey_len)?.to_vec();

        let sig_len_u64 = cursor.read_compact_size()?;
        let sig_len: usize = sig_len_u64
            .try_into()
            .map_err(|_| "parse: sig_len overflows usize".to_string())?;
        let signature = cursor.read_exact(sig_len)?.to_vec();

        witnesses.push(WitnessItem {
            suite_id,
            pubkey,
            signature,
        });
    }

    let da_payload_len_u64 = cursor.read_compact_size()?;
    let da_payload_len: usize = da_payload_len_u64
        .try_into()
        .map_err(|_| "parse: da_payload_len overflows usize".to_string())?;
    let da_payload = cursor.read_exact(da_payload_len)?.to_vec();

    match tx_kind {
        TX_KIND_STANDARD => {
            if da_payload_len != 0 {
                return Err("TX_ERR_PARSE".into());
            }
        }
        TX_KIND_DA_COMMIT => {
            if da_payload_len == 0 || da_payload_len > MAX_DA_MANIFEST_BYTES_PER_TX {
                return Err("TX_ERR_PARSE".into());
            }
        }
        TX_KIND_DA_CHUNK => {
            if da_payload_len == 0 || da_payload_len > MAX_DA_CHUNK_BYTES_PER_TX {
                return Err("TX_ERR_PARSE".into());
            }
        }
        _ => return Err("TX_ERR_PARSE".into()),
    }

    Ok(Tx {
        version,
        tx_kind,
        tx_nonce,
        inputs,
        outputs,
        locktime,
        da_commit,
        da_chunk,
        da_payload,
        witness: WitnessSection { witnesses },
    })
}

pub fn parse_block_bytes(bytes: &[u8]) -> Result<Block, String> {
    let mut cursor = Cursor::new(bytes);
    let header = parse_block_header_from_cursor(&mut cursor)?;
    let tx_count_u64 = cursor.read_compact_size()?;
    let tx_count: usize = tx_count_u64
        .try_into()
        .map_err(|_| "parse: tx_count overflows usize".to_string())?;
    let mut txs = Vec::with_capacity(tx_count);
    for _ in 0..tx_count {
        let tx = parse_tx_from_cursor(&mut cursor)?;
        txs.push(tx);
    }
    if cursor.pos != bytes.len() {
        return Err(BLOCK_ERR_PARSE.into());
    }
    Ok(Block {
        header,
        transactions: txs,
    })
}

fn parse_block_header_from_cursor(cursor: &mut Cursor<'_>) -> Result<BlockHeader, String> {
    let version = cursor.read_u32le()?;
    let prev_block_hash_slice = cursor.read_exact(32)?;
    let mut prev_block_hash = [0u8; 32];
    prev_block_hash.copy_from_slice(prev_block_hash_slice);
    let merkle_root_slice = cursor.read_exact(32)?;
    let mut merkle_root = [0u8; 32];
    merkle_root.copy_from_slice(merkle_root_slice);
    let timestamp = cursor.read_u64le()?;
    let target_slice = cursor.read_exact(32)?;
    let mut target = [0u8; 32];
    target.copy_from_slice(target_slice);
    let nonce = cursor.read_u64le()?;
    Ok(BlockHeader {
        version,
        prev_block_hash,
        merkle_root,
        timestamp,
        target,
        nonce,
    })
}
